ktor-server/ktor-server-plugins/ktor-server-di/common/src/io/ktor/server/plugins/di/DependencyInjection.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2014-2025 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.server.plugins.di
+
+import io.ktor.server.application.ApplicationPlugin
+import io.ktor.server.application.createApplicationPlugin
+import io.ktor.util.AttributeKey
+import io.ktor.util.reflect.TypeInfo
+
+/**
+ * A Ktor application plugin for managing the registration and resolution of module dependencies.
+ *
+ * The following properties are configurable:
+ * 1. `provider`: the logic for registering new types
+ * 2. `resolution`: the function for validating the provided types and creating the dependency map
+ * 3. `reflection`: controls how objects are initialized from type references
+ */
+public val DI: ApplicationPlugin<DependencyInjectionConfig> =
+    createApplicationPlugin("DI", ::DependencyInjectionConfig) {
+        application.attributes.put(
+            DependencyRegistryKey,
+            DependencyRegistryImpl(
+                pluginConfig.provider,
+                pluginConfig.resolution,
+                pluginConfig.reflection,
+            )
+        )
+    }
+
+public val DependencyRegistryKey: AttributeKey<DependencyRegistry> =
+    AttributeKey<DependencyRegistry>("DependencyRegistry")
+
+public expect val DefaultReflection: DependencyReflection
```

Comment:
Probably, we should also provide default "NoReflection" implementation to make it possible to disable reflection.